core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java

```diff
@@ -596,49 +602,75 @@ private void calculateMinTopologyId(int idOfRemovedTransaction) {
       }
    }
 
-   private boolean areTxsOnGoing() {
-      return !localTransactions.isEmpty() || (remoteTransactions != null && !remoteTransactions.isEmpty());
-   }
-
    private void shutDownGracefully() {
       if (log.isDebugEnabled())
          log.debugf("Wait for on-going transactions to finish for %s.", Util.prettyPrintTime(configuration.transaction().cacheStopTimeout(), TimeUnit.MILLISECONDS));
-      // TODO: we can't use TimeService for this since some tests replace the time service and there is a bug
-      // ISPN-5507 that makes it so a tx misses the completion.  When this is combined with a replaced timeService
-      // this will never time out.
+      // We can't use TimeService for this: some tests replace the time service so that it never
+      // advances by cacheStopTimeout milliseconds.
       long now = System.nanoTime();
       long failTime = now + TimeUnit.MILLISECONDS.toNanos(configuration.transaction().cacheStopTimeout());
-      boolean txsOnGoing = areTxsOnGoing();
-      while (txsOnGoing && System.nanoTime() - failTime < 0) {
+
+      boolean localTxsOnGoing = !localTransactions.isEmpty();
+      while (localTxsOnGoing && System.nanoTime() - failTime < 0) {
          try {
             Thread.sleep(30);
-            txsOnGoing = areTxsOnGoing();
+            localTxsOnGoing = !localTransactions.isEmpty();
          } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
-            if (clustered) {
-               log.debugf("Interrupted waiting for on-going transactions to finish. %s local transactions and %s remote transactions", localTransactions.size(), remoteTransactions.size());
-            } else {
-               log.debugf("Interrupted waiting for %s on-going transactions to finish.", localTransactions.size());
+            log.debugf("Interrupted waiting for %d on-going local transactions to finish.", localTransactions.size());
+         }
+      }
+
+      if (executorService != null) {
+         Future<?> remoteTxsFuture = executorService.submit(new Runnable() {
```

Comment:
Yeah, I didn't want a new executor because one thread per cache is already too much here. I guess it would have been better to use one of the global thread pools, I'll make the change.